File: app/voice/speech-gate.ts
// An energy gate with a pre-roll.
//
// Not a voice activity detector: it compares a frame's amplitude against an
// adapting floor and nothing more, and it will happily call a slammed door
// speech. It is deliberately not asked to do more than two jobs, both of which
// tolerate being wrong occasionally:
//
//   Waking. The transcriber bills every second it hears, so it is put to sleep
//   after a long quiet stretch. Something has to decide when the room started
//   making noise again, and replay the half-second before it decided, so the
//   first syllable survives the wait for a session to open.
//
//   Barge-in. While the assistant is talking, `bargeInDetected` — a higher
//   threshold held for longer — is what stops the speaker.
//
// What it explicitly does *not* do is choose which frames go up the wire while
// the upstream is awake. Every frame does, silence included: the transcriber
// finds the end of a turn in the pause after it, and a client that goes quiet
// at the pause never lets it get there.
//
// Pure: no DOM, no timers, no audio API. Frames arrive with the level already
// measured and a clock reading already taken, so the whole policy is testable
// with plain numbers.

/** One frame handed to the gate, and what the gate decided about it. */
export interface SpeechGateResultV1 {
  /** The gate is open: this frame belongs to an utterance. */
  speaking: boolean;
  /** This frame is the one that opened the gate. */
  onset: boolean;
  /** This frame is the one on which the hangover expired. */
  released: boolean;
  /**
   * The pre-roll, in capture order, followed by the current frame. Non-empty
   * only on an onset, and read only when that onset is waking a sleeping
   * upstream — while it is awake the caller sends every frame itself.
   */
  send: ArrayBuffer[];
}

export interface SpeechGateOptionsV1 {
  /** Milliseconds of audio per frame. */
  frameMs: number;
  /** Consecutive loud frames that verify an onset. Default 3 (≈120 ms). */
  onsetFrames?: number;
  /** How long the gate stays open after the last loud frame. Default 900 ms. */
  hangoverMs?: number;
  /** No frame under this RMS is ever speech, however quiet the room. */
  minFloorRms?: number;
  /** How far above the adapted noise floor a frame must sit, in decibels. */
  marginDb?: number;
  /** How much audio the ring holds ahead of an onset. Default 500 ms. */
  prerollMs?: number;
  /** Consecutive loud frames that verify a barge-in. Default 5 (≈200 ms). */
  bargeInFrames?: number;
  /** The barge-in margin, in decibels. Default 6 dB above `marginDb`. */
  bargeInMarginDb?: number;
}

export interface SpeechGateV1 {
  push(level: number, pcm: ArrayBuffer, nowMs: number): SpeechGateResultV1;
  /** How long the gate has been shut. Zero while it is open. */
  quietSinceMs(nowMs: number): number;
  /**
   * The stricter threshold, held for longer, for use while the assistant is
   * talking. It buys a margin against a cough or a keyboard; it does not
   * promise one.
   */
  bargeInDetected(): boolean;
  /** True while the gate is open or inside its hangover. The hangover is what
   * keeps a pause inside a sentence from starting the sleep countdown. */
  open(): boolean;
  /** The adapted noise floor, exposed for tests and diagnostics. */
  noiseFloor(): number;
  /** Forgets the pre-roll and closes the gate — a mute, or a fresh call. */
  reset(nowMs: number): void;
}

const DEFAULT_ONSET_FRAMES = 3;
const DEFAULT_HANGOVER_MS = 900;
const DEFAULT_MIN_FLOOR_RMS = 0.02;
const DEFAULT_MARGIN_DB = 8;
const DEFAULT_PREROLL_MS = 500;
const DEFAULT_BARGE_IN_FRAMES = 5;
const BARGE_IN_EXTRA_DB = 6;

/*
 * The floor climbs very slowly and falls quickly, on purpose, and it does both
 * whether or not the gate is open.
 *
 * Adapting only through silence cannot learn a noise that was already there
 * when the call started: the fan opens the gate on the first frame and holds
 * it open forever, and the meter never stops. Adapting quickly cannot survive
 * a long sentence: the floor chases the speech and shuts the speaker out
 * mid-breath. The rates below are the two answers at once — roughly a
 * ten-second climb and a quarter-second fall.
 *
 * The climb is slower again while the gate is open, because that is where the
 * two requirements pull hardest: a fan the call opened on must still be learnt
 * (it is, in about fifteen seconds, and then the meter stops), and a long
 * unbroken sentence must not be cut off by the floor chasing it (it is not —
 * over half a minute of continuous speech would be needed).
 */
const FLOOR_RISE = 0.004;
const FLOOR_RISE_OPEN = 0.0012;
const FLOOR_FALL = 0.15;

function clampLevel(value: number): number {
  if (!Number.isFinite(value) || value <= 0) return 0;
  return value > 1 ? 1 : value;
}

export function createSpeechGateV1(options: SpeechGateOptionsV1): SpeechGateV1 {
  const frameMs = options.frameMs > 0 ? options.frameMs : 20;
  const onsetFrames = Math.max(1, options.onsetFrames ?? DEFAULT_ONSET_FRAMES);
  const hangoverMs = options.hangoverMs ?? DEFAULT_HANGOVER_MS;
  const minFloor = options.minFloorRms ?? DEFAULT_MIN_FLOOR_RMS;
  const marginDb = options.marginDb ?? DEFAULT_MARGIN_DB;
  const prerollMs = options.prerollMs ?? DEFAULT_PREROLL_MS;
  const bargeInFrames = Math.max(
    1,
    options.bargeInFrames ?? DEFAULT_BARGE_IN_FRAMES,
  );
  const bargeInMarginDb =
    options.bargeInMarginDb ?? marginDb + BARGE_IN_EXTRA_DB;

  const margin = 10 ** (marginDb / 20);
  const bargeInMargin = 10 ** (bargeInMarginDb / 20);
  const ringFrames = Math.max(1, Math.round(prerollMs / frameMs));

  let floor = minFloor;
  let opened = false;
  let loudRun = 0;
  let bargeRun = 0;
  let lastLoudAt = 0;
  let closedSince = 0;
  let started = false;
  // Only frames that were never sent live: emitting the ring on an onset can
  // then never repeat audio the wire has already carried.
  const ring: ArrayBuffer[] = [];

  function threshold(): number {
    return Math.max(minFloor, floor * margin);
  }

  function bargeThreshold(): number {
    return Math.max(minFloor * bargeInMargin, floor * bargeInMargin);
  }

  return {
    push(rawLevel, pcm, nowMs) {
      const level = clampLevel(rawLevel);
      if (!started) {
        started = true;
        closedSince = nowMs;
      }
      const loud = level >= threshold();
      loudRun = loud ? loudRun + 1 : 0;
      bargeRun = level >= bargeThreshold() ? bargeRun + 1 : 0;

      const rate =
        level <= floor ? FLOOR_FALL : opened ? FLOOR_RISE_OPEN : FLOOR_RISE;
      floor += (level - floor) * rate;

      let onset = false;
      let released = false;
      const send: ArrayBuffer[] = [];

      if (!opened && loudRun >= onsetFrames) {
        opened = true;
        onset = true;
        lastLoudAt = nowMs;
        send.push(...ring.splice(0, ring.length), pcm);
        return { speaking: true, onset, released, send };
      }

      if (opened) {
        if (loud) lastLoudAt = nowMs;
        if (nowMs - lastLoudAt > hangoverMs) {
          opened = false;
          released = true;
          closedSince = nowMs;
          loudRun = 0;
        } else {
          send.push(pcm);
          return { speaking: true, onset, released, send };
        }
      }

      ring.push(pcm);
      while (ring.length > ringFrames) ring.shift();
      return { speaking: false, onset, released, send };
    },

    quietSinceMs(nowMs) {
      if (opened) return 0;
      if (!started) return 0;
      return Math.max(0, nowMs - closedSince);
    },

    bargeInDetected() {
      return bargeRun >= bargeInFrames;
    },

    open() {
      return opened;
    },

    noiseFloor() {
      return floor;
    },

    reset(nowMs) {
      opened = false;
      loudRun = 0;
      bargeRun = 0;
      lastLoudAt = nowMs;
      closedSince = nowMs;
      started = true;
      ring.length = 0;
    },
  };
}

File: apps/native/lib/voice/speech_gate.dart
/// A speech gate with a pre-roll: when someone started talking, and when the
/// room has been quiet long enough to stop paying for it.
///
/// Timing is this object's whole job — onset, hangover, pre-roll. The label
/// "this frame is speech" can come from energy (the default: an adaptive
/// floor) or from a speech probability the caller already computed, which is
/// how Silero replaces the slammed-door failure without owning the clock.
/// Either way this is not a turn detector. Nothing above it may assume noise
/// has been rejected from the audio the upstream hears.
///
/// It decides exactly two things, and the caller decides the rest:
///
/// * **Wake.** A sleeping upstream is woken by a verified onset, and the
///   pre-roll ring is replayed ahead of the live audio so the first syllable
///   is not the price of waking up.
/// * **Barge-in.** Verified speech by a stricter margin, while the assistant
///   is speaking, is offered as a reason to stop it — the controller takes it
///   only for a capture whose playback echo it can cancel.
///
/// What it must never be used for is deciding whether an individual frame is
/// worth sending once the upstream is awake. The server's transcriber decides
/// where a turn ends and needs the silence after the words to decide it —
/// about half a second — so an awake session gets a frame every 40 ms (on a
/// capture that cannot cancel its own playback, silence stands in for the
/// microphone while the reply plays; that is the controller's rule, not this
/// gate's). [quietForMs] exists for the one policy that does stop the audio:
/// two minutes of quiet while listening.
///
/// Pure Dart over a level, an optional probability and a timestamp: no audio
/// API, no timers, no clock of its own. The caller supplies the timestamps,
/// which is what lets a test run two minutes of silence in a millisecond.
library;

import 'dart:collection';
import 'dart:math' as math;
import 'dart:typed_data';

/// What a room at rest sounds like to a working microphone, and so the line
/// between a device that is hearing and one handing over nothing.
///
/// A microphone picks up a room — a fan, the building, the device's own noise
/// — so this is far below anything the gate would call speech: it is not
/// [SpeechGateConfig.floor], which is a threshold for words. Only a device
/// delivering zeros sits under it.
const voiceRoomToneLevelV1 = 0.0005;

/// The RMS of one PCM16 little-endian frame, 0..1.
///
/// An odd trailing byte is ignored rather than read as half a sample.
double pcm16Rms(Uint8List bytes) {
  final samples = bytes.lengthInBytes ~/ 2;
  if (samples == 0) return 0;
  final view = ByteData.sublistView(bytes, 0, samples * 2);
  var sum = 0.0;
  for (var i = 0; i < samples; i++) {
    final sample = view.getInt16(i * 2, Endian.little) / 32768.0;
    sum += sample * sample;
  }
  return math.sqrt(sum / samples);
}

class SpeechGateConfig {
  /// How long one frame of audio is. Everything below is counted in frames.
  final Duration frame;

  /// Energy must hold above the floor for this long before the gate opens.
  final Duration onset;

  /// The gate stays open this long after the last loud frame, so a breath
  /// mid-sentence does not read as the end of one.
  final Duration hangover;

  /// How much audio is replayed ahead of an onset.
  final Duration preRoll;

  /// How far above the noise floor a frame must be to count as speech.
  final double onsetMargin;

  /// The stricter margin used while the assistant is speaking.
  final double bargeInMargin;

  /// Barge-in needs verified speech for this long.
  final Duration bargeInHold;

  /// Nothing quieter than this is speech, however quiet the room is.
  final double floor;

  /// Speech probability at or above this is an onset, when the caller is
  /// labelling frames (Silero) rather than using energy.
  final double speechThreshold;

  /// The stricter probability used for barge-in with a labelled frame.
  final double bargeSpeechThreshold;

  const SpeechGateConfig({
    this.frame = const Duration(milliseconds: 40),
    this.onset = const Duration(milliseconds: 120),
    this.hangover = const Duration(milliseconds: 900),
    this.preRoll = const Duration(milliseconds: 500),
    this.onsetMargin = 2.5,
    this.bargeInMargin = 4.5,
    this.bargeInHold = const Duration(milliseconds: 200),
    this.floor = 0.012,
    this.speechThreshold = 0.5,
    this.bargeSpeechThreshold = 0.7,
  });

  int get onsetFrames =>
      math.max(1, (onset.inMicroseconds / frame.inMicroseconds).ceil());
  int get preRollFrames =>
      math.max(1, (preRoll.inMicroseconds / frame.inMicroseconds).ceil());
  int get bargeInFrames =>
      math.max(1, (bargeInHold.inMicroseconds / frame.inMicroseconds).ceil());
}

/// What the caller should do with the frame it just offered.
class SpeechGateDecision {
  /// The frames to send, oldest first. On an onset this is the pre-roll ring
  /// ending with the frame just offered; while open it is that frame alone;
  /// while closed it is empty.
  final List<Uint8List> emit;

  /// Whether the gate is open — or inside its hangover, which is the same
  /// thing to a caller.
  final bool open;

  /// This frame is the one that opened the gate.
  final bool onset;

  /// This frame is the one that closed it.
  final bool closed;

  /// Verified speech by the stricter margin: enough to interrupt a reply.
  final bool bargeIn;

  const SpeechGateDecision({
    required this.emit,
    required this.open,
    required this.onset,
    required this.closed,
    required this.bargeIn,
  });
}

class SpeechGate {
  final SpeechGateConfig config;
  final ListQueue<Uint8List> _ring = ListQueue<Uint8List>();
  double _noiseFloor = 0;
  int _adapted = 0;
  int _consecutive = 0;
  int _bargeConsecutive = 0;
  bool _open = false;
  int _lastVoiceAt = 0;
  int _quietSince = 0;
  int _lastAt = 0;

  SpeechGate({this.config = const SpeechGateConfig(), int startedAtMs = 0})
    : _quietSince = startedAtMs,
      _lastAt = startedAtMs;

  bool get open => _open;

  /// When the gate last closed, in the caller's own milliseconds. While the
  /// gate is open this is the moment it opened, which no caller reads.
  int get quietSince => _quietSince;

  /// How long the gate has been closed at [nowMs]; zero while it is open.
  int quietForMs(int nowMs) => _open ? 0 : math.max(0, nowMs - _quietSince);

  /// The adaptive noise floor, exposed for tests and diagnostics.
  double get noiseFloor => _noiseFloor;

  double get threshold =>
      math.max(_noiseFloor * config.onsetMargin, config.floor);

  /// Offers one frame and its level. [atMs] is the caller's clock.
  ///
  /// [speechScore] is a speech probability 0..1. When it is set, onset and
  /// barge-in read it instead of energy — that is the Silero path. The energy
  /// floor is left alone on that path: a slammed door is loud and not speech,
  /// and must not train the fallback.
  SpeechGateDecision offer(
    Uint8List frame,
    double level,
    int atMs, {
    double? speechScore,
  }) {
    _lastAt = atMs;
    final loud = speechScore != null
        ? speechScore >= config.speechThreshold
        : level >= threshold;
    // The floor only learns from quiet: adapting to speech would raise the
    // bar until nothing is speech.
    if (speechScore == null && !loud) {
      final alpha = _adapted < 10 ? 0.25 : 0.05;
      _noiseFloor = _noiseFloor * (1 - alpha) + level * alpha;
      _adapted++;
    }
    _consecutive = loud ? _consecutive + 1 : 0;

    final bargeLoud = speechScore != null
        ? speechScore >= config.bargeSpeechThreshold
        : level >=
              math.max(_noiseFloor * config.bargeInMargin, config.floor * 2);
    _bargeConsecutive = bargeLoud ? _bargeConsecutive + 1 : 0;
    final bargeIn = _bargeConsecutive >= config.bargeInFrames;

    _ring.addLast(frame);
    while (_ring.length > config.preRollFrames) {
      _ring.removeFirst();
    }

    var onset = false;
    var closed = false;
    if (_open) {
      if (loud) _lastVoiceAt = atMs;
      if (atMs - _lastVoiceAt >= config.hangover.inMilliseconds) {
        _open = false;
        closed = true;
        _quietSince = atMs;
      }
    } else if (_consecutive >= config.onsetFrames) {
      _open = true;
      onset = true;
      _lastVoiceAt = atMs;
    }

    // Draining on every emitted frame is what keeps the ring from ever
    // sending the same frame twice: what leaves is gone. The frame that
    // closes the gate is a hangover past the last word, so it stays in the
    // ring as pre-roll for whatever is said next.
    final emit = <Uint8List>[];
    if (_open) {
      emit.addAll(_ring);
      _ring.clear();
    }
    return SpeechGateDecision(
      emit: emit,
      open: _open,
      onset: onset,
      closed: closed,
      bargeIn: bargeIn,
    );
  }

  /// The pre-roll as it stands, oldest first, without disturbing the gate.
  /// This is what a wake replays when the onset itself was not the drain.
  List<Uint8List> preRoll() => List<Uint8List>.of(_ring);

  /// Drops the pre-roll: what the caller has already sent must not be sent
  /// again on the next onset.
  void clearPreRoll() => _ring.clear();

  /// Forgets the conversation but keeps the learned noise floor: a mute or a
  /// sleep is not a new room.
  void reset() {
    _ring.clear();
    _consecutive = 0;
    _bargeConsecutive = 0;
    _open = false;
    _quietSince = _lastAt;
  }
}
